runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -1026,4 +1019,31 @@ private boolean isAnnotationPresent(AnnotatedElement annotatedElement,
 		cache.put(annotatedElement, isPresent);
 		return isPresent;
 	}
+
+	/**
+	 * The IdentityWeakReference extends the {@link WeakReference} with the
+	 * difference that it can be compared to another IdentityWeakReference. The
+	 * compare is done on the Identity of the Referenced Object, this allows us to
+	 * use this element in a Set and have unique objects in it.
+	 */
+	private static class IdentityWeakReference<T> extends WeakReference<T> {
+
+		IdentityWeakReference(T referent) {
+			super(referent);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj instanceof IdentityWeakReference other) {
+				return Objects.equals(other.get(), get());
```

Comment:
The existing implementation was using == but here you using .equals.  Also, this definition of equals is not consistent with the hashCode implementation which uses the identity hash code.  I think you should simply use == here given this whole class is about "identity".

Reply:
Yes thats probabbly better here.

Reply:
You are right, I did take a look into Objects equals, but this does both checks, == and equals